sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -94,11 +178,95 @@ trait ScalaReflection {
     case (other, _) => other
   }
 
+  /**
+   * Creates a converter function that will convert Catalyst types to Scala type.
+   */
+  private[sql] def createScalaConverter(dataType: DataType): Any => Any = dataType match {
+    // Check UDT first since UDTs can override other types
+    case udt: UserDefinedType[_] =>
+      (item: Any) => udt.deserialize(item)
+
+    case arrayType: ArrayType =>
+      val elementConverter = createScalaConverter(arrayType.elementType)
+      (item: Any) => item match {
+        case s: Seq[_] => s.map(elementConverter)
+        case other => other
```

Comment:
Is this needed? It may lead to other problems (the items inside other is not converted).